### PR TITLE
Gradle check retry

### PR DESCRIPTION
**Pull request by kotwanikunal**
### Description
- Adds a retry mechanism for all test classes under the root project
- This should help with `gradle check` failures
 
### Issues Resolved
- https://github.com/opensearch-project/OpenSearch/issues/2547
 
### Check List
- [ ] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -386,6 +387,20 @@ gradle.projectsEvaluated {
   }
 }
 
+// test retry configuration
+boolean isCiServer = System.getenv().containsKey("CI")
+subprojects {
+  apply plugin: "org.gradle.test-retry"
+  tasks.withType(Test).configureEach {
+    retry {
+      if (isCiServer) {
```
Comment (reta) [inline build.gradle:396]: You could probably just not configure the task if it is not CI:

```
if (isCiServer) {
tasks.withType(Test).configureEach {
    retry {
     ....
    }
}
```

Comment (reta) [reply build.gradle:396]: Also, I would highly recommend to set `maxFailures` to some reasonable value, `5` for example:

```
     * The maximum number of test failures that are allowed before retrying is disabled.
```

Wdyt?

**Comment by dblock** [comment]
The fact that gradle check succeeded here from the first try tells me that we're on the right track :)

File: build.gradle (hunk 1 above)
Comment (kotwanikunal) [reply build.gradle:396]: We would really benefit from something like https://github.com/gradle/test-retry-gradle-plugin/issues/29 for our use case to catch the worst offenders.
I agree with giving it a kick start with 1-2 retries and stopping on multiple test failures. From what I have noticed, it usually takes 2 retries recently to get it going. But let's start small enough to give a boost and not over rely.

I'll update the PR with `{ maxFailures = 10, maxRetries = 1 }`.

**Comment by kotwanikunal** [comment]
```
> Task :client:rest-high-level:asyncIntegTest

REPRODUCE WITH: ./gradlew ':client:rest-high-level:asyncIntegTest' --tests "org.opensearch.client.TasksIT.testGetValidTask" -Dtests.seed=91F1364AFC43A70 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=sr-ME -Dtests.timezone=Pacific/Enderbury -Druntime.java=17

org.opensearch.client.TasksIT > testGetValidTask FAILED
    java.lang.AssertionError
        at __randomizedtesting.SeedInfo.seed([91F1364AFC43A70:9573CA83C6999BB]:0)
        at org.junit.Assert.fail(Assert.java:87)
        at org.junit.Assert.assertTrue(Assert.java:42)
        at org.junit.Assert.assertTrue(Assert.java:53)
        at org.opensearch.client.OpenSearchRestHighLevelClientTestCase.lambda$checkTaskCompletionStatus$2(OpenSearchRestHighLevelClientTestCase.java:361)
        at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1060)
        at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1033)
        at org.opensearch.client.TasksIT.testGetValidTask(TasksIT.java:123)
```

**Comment by kotwanikunal** [comment]
start gradle check

**Comment by dblock** [comment]
@kotwanikunal Did it retry `asyncIntegTest`? Where do you see it in the log?

**Comment by reta** [comment]
@kotwanikunal @dblock I believe the `asyncIntegTest` may not fall into `Test` task type:

```
tasks.withType(Test).configureEach {
    ...
}
```

We do some tricks here as well https://github.com/opensearch-project/OpenSearch/blob/main/gradle/code-coverage.gradle#L47

**Comment by kotwanikunal** [comment]
> @kotwanikunal Did it retry `org.opensearch.client.TasksIT.testGetValidTask` or `asyncIntegTest`? Where do you see it in the log?

It only retries the failing tests. I verified that on my local machine with different scenarios including `integTest`, `internalClusterTest`

In the individual project logs, you usually see `x tests completed, n tests failed` with the html report where n counts all the retries.

I will try force failing an `asyncIntegTest` on my devbox just to be sure.

**Comment by kotwanikunal** [comment]
> @kotwanikunal @dblock I believe the `asyncIntegTest` may not fall into `Test` task type:
> 
> ```
> tasks.withType(Test).configureEach {
>     ...
> }
> ```
> 
> We do some tricks here as well https://github.com/opensearch-project/OpenSearch/blob/main/gradle/code-coverage.gradle#L47

It does infact cover the `asyncIntegTest` type.

I ran `./gradlew ':client:rest-high-level:asyncIntegTest' --tests "org.opensearch.client.TasksIT.testGetValidTask" -Dtests.seed=91F1364AFC43A70 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=sr-ME -Dtests.timezone=Pacific/Enderbury -Druntime.java=17`  on my devbox by adding in a failing assert.

It looks like it retries the failing `asyncIntegTest` type - 

image

**Review by dblock — APPROVED** [review]
This is great!

**Comment by tlfeng** [comment]
About `asyncIntegTest` task, seems it's under `Test` type.

Its direct type is `RestIntegTestTask` https://github.com/opensearch-project/OpenSearch/blob/1.3.0/client/rest-high-level/build.gradle#L84.
Then `RestIntegTestTask` extends `StandaloneRestIntegTestTask`, which extends `Test`.
https://github.com/opensearch-project/OpenSearch/blob/1.3.0/buildSrc/src/main/java/org/opensearch/gradle/test/RestIntegTestTask.java#L44
https://github.com/opensearch-project/OpenSearch/blob/1.3.0/buildSrc/src/main/java/org/opensearch/gradle/testclusters/StandaloneRestIntegTestTask.java#L61

**Comment by tlfeng** [comment]
What wired here is I didn't see the retry occurs in the above failed gradle check, in the "report 3865"
image

**Comment by dblock** [comment]
> What wired here is I didn't see the retry occurs in the above failed gradle check, in the "report 3865" image

Is `boolean isCiServer = System.getenv().containsKey("CI")` failing here? Maybe just put retry all the time? It's useful for dev too.

**Comment by kotwanikunal** [comment]
> > What wired here is I didn't see the retry occurs in the above failed gradle check, in the "report 3865" image
> 
> Is `boolean isCiServer = System.getenv().containsKey("CI")` failing here? Maybe just put retry all the time? It's useful for dev too.

Looks like it. Since it's a single retry, let me get rid of the CI condition. I can iteratively improve on it if we decide to have different configs for dev/CI.

**Comment by reta** [comment]
@kotwanikunal    use `BuildParams` instead?

```     
if (BuildParams.isCi() == true) {
}
```

**Comment by kotwanikunal** [comment]
> @kotwanikunal use `BuildParams` instead?
> 
> ```
> if (BuildParams.isCi() == true) {
> }
> ```

@reta How about adding in a single retry for dev and CI for now?

**Comment by dblock** [comment]
> > @kotwanikunal use `BuildParams` instead?
> > ```
> > if (BuildParams.isCi() == true) {
> > }
> > ```
> 
> @reta How about adding in a single retry for dev and CI for now?

Let's vote and if we have a deadlock, we'll just retry? :)

**Comment by reta** [comment]
> Let's vote and if we have a deadlock, we'll just try again? :) I'm good with anything.

I am fine with that ... for now

**Comment by mch2** [comment]
How do we ensure we aren't swallowing failures with reproducible seeds? Does it retry with the same seed?

**Comment by reta** [comment]
> How do we ensure we aren't swallowing failures with reproducible seeds? Does it retry with the same seed?

I am not 100% sure but I would doubt the seed is preserved

**Comment by reta** [comment]
> 

Aha, this one seems to retried 

image

And same seed was used 

```
testHAProxyModeConnectionWorks
java.lang.AssertionError
	at __randomizedtesting.SeedInfo.seed([108B309DF15FC962:17FDE556CAD9BF3F]:0)
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertTrue(Assert.java:53)
```

```
java.lang.AssertionError
	at __randomizedtesting.SeedInfo.seed([108B309DF15FC962:17FDE556CAD9BF3F]:0)
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertTrue(Assert.java:53)
	at org.opensearch.cluster.remote.test.RemoteClustersIT.testHAProxyModeConnectionWorks(RemoteClustersIT.java:125)
	at java.base/jdk.internal.
```

@mch2 ^^

**Comment by kotwanikunal** [comment]
Open issue for this test: https://github.com/opensearch-project/OpenSearch/issues/1703

**Comment by kotwanikunal** [comment]
> 



> > 
> 
> Aha, this one seems to retried
> 
> image
> 
> And same seed was used
> 
> ```
> testHAProxyModeConnectionWorks
> java.lang.AssertionError
> 	at __randomizedtesting.SeedInfo.seed([108B309DF15FC962:17FDE556CAD9BF3F]:0)
> 	at org.junit.Assert.fail(Assert.java:87)
> 	at org.junit.Assert.assertTrue(Assert.java:42)
> 	at org.junit.Assert.assertTrue(Assert.java:53)
> ```
> 
> ```
> java.lang.AssertionError
> 	at __randomizedtesting.SeedInfo.seed([108B309DF15FC962:17FDE556CAD9BF3F]:0)
> 	at org.junit.Assert.fail(Assert.java:87)
> 	at org.junit.Assert.assertTrue(Assert.java:42)
> 	at org.junit.Assert.assertTrue(Assert.java:53)
> 	at org.opensearch.cluster.remote.test.RemoteClustersIT.testHAProxyModeConnectionWorks(RemoteClustersIT.java:125)
> 	at java.base/jdk.internal.
> ```
> 
> @mch2 ^^
> 
> @dblock seems like 1 retry is not enough ... or we need to retry the while suite?

Retrying the whole suite will make it difficult to pinpoint flakiness to particular tests. This helps us get to the most frequent offenders IMHO.
But I am all ears about what y'all have to say 🙂

**Comment by reta** [comment]
> Retrying the whole suite will make it difficult to pinpoint flakiness to particular tests. This helps us get to the most frequent offenders IMHO.

I agree, I don't like that at all, thinking loudly - if cluster ended up in "bad" ( == not expected) state at startup, the test may never succeed ... but we should consider that to be separate problem I guess

**Comment by dblock** [comment]
Do you want to increase the number of retries @kotwanikunal?

**Comment by tlfeng** [comment]
@kotwanikunal Would you backport the change into 1.x branch?

**Pull request merged**